sympy/stats/stochastic_process_types.py

```diff
@@ -0,0 +1,315 @@
+    """
+    def __getitem__(self, time):
+        """
+        For indexing discrete time stochastic processes.
+
+        Returns
+        =======
+
+        RandomIndexedSymbol
+        """
+        if time not in self.index_set:
+            raise IndexError("%s is not in the index set of %s"%(time, self.symbol))
+        idx_obj = Indexed(self.symbol, time)
+        pspace_obj = StochasticPSpace(self.symbol, self)
+        return RandomIndexedSymbol(idx_obj, pspace_obj)
+
+# TODO: Define ContinuousTimeStochasticProcess
+
+class TransitionMatrix(Boolean):
+    """
+    Assumes that the matrix is the transition matrix
+    of the process.
+    """
+
+    def __new__(cls, process, matrix):
+        if not isinstance(process, DiscreteMarkovChain):
+            raise ValueError("Currently only DiscreteMarkovChain "
+                                "support TransitionMatrix.")
+        matrix = _matrix_checks(matrix)
+        return Basic.__new__(cls, process, matrix)
+
+    process = property(lambda self: self.args[0])
+    matrix = property(lambda self: self.args[1])
+
+class StateSpace(Boolean):
+
+    def __new__(cls, process, state_space):
+        if not isinstance(process, DiscreteMarkovChain):
+            raise ValueError("Currently only DiscreteMarkovChain "
+                                "support StateSpace.")
+        state_space = _set_converter(state_space)
+        return Basic.__new__(cls, process, state_space)
+
+    process = property(lambda self: self.args[0])
+    state_space = property(lambda self: self.args[1])
+
+class DiscreteMarkovChain(DiscreteTimeStochasticProcess):
+    """
+    Represents discrete Markov chain.
+
+    Parameters
+    ==========
+
+    sym: Symbol
+    state_space: Set
+        Optional, by default, S.Reals
+    trans_probs: Matrix/ImmutableMatrix/MatrixSymbol
+        Optional, by default, None
+
+    Examples
+    ========
+
+    >>> from sympy.stats import DiscreteMarkovChain, TransitionMatrix
+    >>> from sympy import Matrix, MatrixSymbol, Eq
+    >>> from sympy.stats import P
+    >>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
+    >>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
+    >>> YS = DiscreteMarkovChain("Y")
+    >>> Y.state_space
+    {0, 1, 2}
+    >>> Y.trans_probs
+    Matrix([
+    [0.5, 0.2, 0.3],
+    [0.2, 0.5, 0.3],
+    [0.2, 0.3, 0.5]])
+    >>> TS = MatrixSymbol('T', 3, 3)
+    >>> P(Eq(YS[3], 2), Eq(YS[1], 1) & TransitionMatrix(YS, TS))
+    T[0, 2]*T[1, 0] + T[1, 1]*T[1, 2] + T[1, 2]*T[2, 2]
+    >>> P(Eq(Y[3], 2), Eq(Y[1], 1)).round(2)
+    0.36
+    """
+
+    index_set = S.Naturals0
+
+    def __new__(cls, sym, state_space=S.Reals, trans_probs=None):
+        sym = _symbol_converter(sym)
+        state_space = _set_converter(state_space)
+        if trans_probs != None:
+            trans_probs = _matrix_checks(trans_probs)
+        return Basic.__new__(cls, sym, state_space, trans_probs)
+
+    @property
+    def trans_probs(self):
```

Comment:
why not transition_probabilities ? I don't see the point in using abbreviations since everyone's using TAB-completion today.